src/libraries/System.Net.Security/tests/FunctionalTests/TestHelper.cs

```diff
@@ -47,5 +48,137 @@ internal static (VirtualNetworkStream ClientStream, VirtualNetworkStream ServerS
+        {
+            X509Certificate2Collection chain = new X509Certificate2Collection();
+
+            using (RSA root = RSA.Create())
+            using (RSA intermediate = RSA.Create())
+            using (RSA server = RSA.Create())
+            {
+                CertificateRequest rootReq = new CertificateRequest(
+                    "CN=Root",
+                    root,
+                    HashAlgorithmName.SHA256,
+                    RSASignaturePadding.Pkcs1);
+
+                rootReq.CertificateExtensions.Add(
+                    new X509BasicConstraintsExtension(true, false, 0, true));
+                rootReq.CertificateExtensions.Add(
+                    new X509SubjectKeyIdentifierExtension(rootReq.PublicKey, false));
+                rootReq.CertificateExtensions.Add(
+                    new X509KeyUsageExtension(X509KeyUsageFlags.KeyCertSign, false));
+
+                //DateTimeOffset start = DateTimeOffset.UtcNow.AddMinutes(-5);
+                DateTimeOffset endTime = startTime.AddMonths(1);
+
+                X509Certificate2 rootCertWithKey = rootReq.CreateSelfSigned(startTime, endTime);
+
+                CertificateRequest intermedReq = new CertificateRequest(
+                   "CN=Intermediate",
+                   intermediate,
+                   HashAlgorithmName.SHA256,
+                   RSASignaturePadding.Pkcs1);
+
+                intermedReq.CertificateExtensions.Add(
+                    new X509BasicConstraintsExtension(true, false, 0, true));
+                intermedReq.CertificateExtensions.Add(
+                    new X509SubjectKeyIdentifierExtension(intermedReq.PublicKey, false));
+                intermedReq.CertificateExtensions.Add(
+                    new X509KeyUsageExtension(X509KeyUsageFlags.KeyCertSign, false));
+
+                byte[] serial = new byte[8];
+                RandomNumberGenerator.Fill(serial);
+
+                X509Certificate2 intermedCertWithKey;
+                using (X509Certificate2 intermedPub = intermedReq.Create(rootCertWithKey, startTime, endTime, serial))
+                {
+                    intermedCertWithKey = intermedPub.CopyWithPrivateKey(intermediate);
+                }
+
+                CertificateRequest serverReq = new CertificateRequest(
+                     $"CN={name}",
+                     server,
+                     HashAlgorithmName.SHA256,
+                     RSASignaturePadding.Pkcs1);
+
+                serverReq.CertificateExtensions.Add(
+                    new X509BasicConstraintsExtension(false, false, 0, false));
+                serverReq.CertificateExtensions.Add(
+                    new X509SubjectKeyIdentifierExtension(serverReq.PublicKey, false));
+
+                // Add Issuer KeyIdentifier
+                using (SHA1 sha1 = SHA1.Create())
+                {
+                    byte[] data = new byte[24];
+                    data[0] = 0x30; //SEQUENCE
```

Comment:
Instead of bit-banging you can use the System.Formats.Asn1.AsnWriter to do this for you.  A lot of examples are in https://github.com/dotnet/runtime/blob/master/src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/CertificateAuthority.cs

Reply:
The AsnWriter is nice. I'm wondering if it would make sense to move some parts to common place and share it. Common/tests/System/Security/Cryptography/CertificateAuthority.cs perhaps?

Reply:
Yeah, moving the CertificateAuthority to Common/tests/System/Security/Cryptography/ seems reasonable.  If it's not straightforward, lemme know and I'll try to get it standalone in the morning.